readme.txt
TODO:
1 迷你地图显示

使用:
丢interface/addon文件夹 左键物品聊天框po出链接

changelog.txt
0.1 第一个版本,实现从handnotes导入的数据 占用79k
0.2 支持物品链接,左键输入链接 占用84k
0.3 调整数据库结构 子地图问题修复 数据库导入 如果是物品id直接用物品材质 占用100k
0.4 更新数据库 清理部分不掉落玩具的稀有 82k
0.5 添加宠物对战 继续瘦身 70k
0.6 完成检测功能 已经完成的半透明 加入山洞坐标 待测试 73k
0.61 测试完毕 修正数据中
0.62 加了个库用于获取地图比例偏移
0.7 折腾了数据库 实现上级地图显示 需要在该组数据表内加level = 2, 占用91k
_0.8 git真tm脆弱 又玩坏了 也没备份 慢慢补回来吧
0.8 优化了逻辑

Date.lua
local _,ns = ...
ns.db = {}

local GetAchievementCriteriaInfo = GetAchievementCriteriaInfo

local RichOfPandaria =  { -- =潘达利亚的宝藏(金币)
	[862] = { --潘达利亚
		["船载储物箱"] = {
			coord = 70137450,
			quest = 31396,
			desc = "沉船船底",
		},
	},
	[807] = { --四风
		["[ 兔妖宝箱 ]"] = {
			["desc"] = "(23.72,28.32)入口隐藏在灌木丛中。",
			["icon"] = 17,
			["coord"] =23253067, -- [1]
			quest=31405,
		},
		["一箱被偷的货物"] = {
			["desc"] = "在汇风岭的兔子洞里，宝箱，随机附魔项链",
			["coord"] =43953808, -- [1]
			quest=31406,
		},
	},
	[809] = { --崑莱
		["失窃的林精宝箱"] = {
			["desc"] = "凛风洞深处(坐标：31.9 50.4)  ",
			["icon"] = 17,
			["coord"] =59405300, --洞口
			redir = "HowlingwindCavern",
		},
		["遗失的冒险者的遗产"] = {
			["desc"] = "废弃营地  ",
			["coord"] =36707980, -- [1]
			quest=31418,
		},
		["墨墨的宝箱"] = {
			["desc"] = "房间里",
			["icon"] = 17,
			["coord"] =48107300, -- [1]
			quest=31868,
		},
		["精灵的衣箱山洞"] = {--"林精的衣箱"
			["desc"] = "凛风洞深处，随机附魔披风，56.3 71.5",
			["icon"] = 17,
			["coord"] =73107350, -- [1]
			redir = "PrankstersHollow"
		},
		["猢狲宝箱"] = {
			["desc"] = "肘锤洞任务怪[邪恶的]裘裘背后 100金 ",
			["icon"] = 17,
			["coord"] =50306180, -- [1]
			redir = "KnucklethumpHole",
		},
		[86471] = {
			["desc"] = "走道非洞穴 右侧雕像背后 ",
			["icon"] = 17,
			["coord"] =64264516, -- [1]
			quest=31420,
		},
		[86422] = {
			["desc"] = "水潭底部",
			["coord"] =72063388, -- [1]
			quest=31416
		},
		[86427] = { 
			["coord"] =57807630, -- [1]
			quest=31422,
		},
	},
	[806] = { --翠玉
		[85780] = {
			["coord"] =26293181, -- [1]
			quest=31400,
		},
		["神龛"] = {
			["desc"] = "被斥责 ",
			["coord"] =46408070, -- [1]
			quest=31865,
		},
		[85781] = {
			["desc"] = "在铁腕机坪边上的一口井里。",
			["coord"] =31962775, -- [1]
			quest=31401, 
		},
		["秘藏的宝藏"] = {--???
			["desc"] = "临风岛山洞(坐标：64 ,25)",
			["icon"] = 17,
			["coord"] =64042510, -- [1]
			quest=31866,
		},
		["船载储物箱"] = {--??
			["desc"] = "翡翠林地图边缘 (坐标：51.4 100.0)  沉没在南部海岸的联盟船只造成的",
			["icon"] = 17,
			["coord"] =51429998, -- [1]
			quest=31396,
		},
		[86216] = {
			["desc"] = "努卡努卡营地西北边,玉盘湖中央。",
			["coord"] =23595108, -- [1]
			quest=31404, 
		},
	},
	[873] = { --客栈
		["被遗忘的密码箱"] = {
			["desc"] = "客栈二楼 10金 ",
			["icon"] = 17,
			["coord"] =55377114, -- [1]
			 quest=31867,
		},
		[86473] = {
			["desc"] = "石匠之愚山顶，黑市的东南边 ",
			["coord"] =75247634, -- [1]
			quest=31428,
		},
	},
	[810] = { --螳螂
		["一箱被丢弃的货物*"] = {
			["desc"] = "蟠龙脊内燃烧的帐篷 (坐标：62.8 34.0)",
			["coord"] =62743440, -- [1]
			quest=31427,
		},
		[86517] = {
			["desc"] = "科里维斯根部",
			["coord"] =53835991, -- [1]
			quest=31424,
		},
		[86472] = {
			["desc"] = "黯光枯木树根下(坐标：65.8 86.0) ",
			["coord"] = 65808610, -- [1]
			quest=31426, 
		},
		[86516] = { --山洞
			["desc"] = "砮皂寺西边的山洞 (坐标：48.4 89.2)",
			["icon"] = 17,
			["coord"] =32766160, -- [1]
			redir = "NiuzaoCatacombs"
		},
	},
	[857] = { --蛮荒
		[87266] = {
			["desc"] = "刷新稀有猴子的船头。",
			["coord"] =52318844, -- [1]
			quest=31411,
		},
		[86220] = {
			["desc"] = "朱家堡西边山洞(坐标：75.0 55.1) *由玩家提供的位置 100金 ",
			["icon"] = 17,
			["coord"] =71180862, -- [1]
			quest=31408,
		},
		["装备箱+"] = {
			["desc"] = "船舱底层，宝箱，随机附魔鞋子",
			["coord"] =41949262, -- [1]
			quest=31410,
		},
	},
	["PrankstersHollow"] = { --崑莱山洞
		["林精的衣箱"] = {
			["desc"] = "凛风洞深处，随机附魔披风，56.3 71.5",
			quest=31412, 
			coord = 54706980,
		},
	},
	["KnucklethumpHole"] = { -- 崑莱山洞
		["猢狲宝箱的洞口"] = {
			coord = 51892701,
			["desc"] = "肘锤洞任务怪[邪恶的]裘裘背后 100金 ",
			quest = 31414,
		},
    },
	["HowlingwindCavern"] = { -- 崑莱山洞
		["失窃的林精宝箱"] = {
			coord = 41404420,
			["desc"] = "凛风洞深处(坐标：31.9 50.4)  ",
			quest=31415, 
		},
    },
	["NiuzaoCatacombs"] = { -- 螳螂山洞
		[86516] = {
			coord = {56406480,36908760,48408860,64502150},
			["desc"] = "砮皂寺西边的山洞 (坐标：48.4 89.2)",
			quest=31423, 
		},
	},
}

local LostAndFound = { -- =装备
	[807] = { --四风
		[86218] = {
			["desc"] = "竹子丛里。非武僧职业可以获得的唯一武僧风格法杖。",
			["coord"] = {
				15402920, -- [1]
				14903360, -- [2]
				17503570, -- [3]
				19103780, -- [4]
				19004250, -- [5]
			},
			quest = 86218,
		},
		[87524] = {
			["coord"] = 92153889, 
			quest=31869,
		},
		[86079] = {
			["coord"] = 45403802, 
			 quest=31292,
		},
		[85973] = {
			["coord"] = 46642476, -- [1]
			quest=31284,
		},
	},
	[809] = { --崑莱
		[86125] = {
			["desc"] = "洞穴最深处,与冰块中的npc64227冰封小径运货人对话获得。",
			["icon"] = 17,
			["coord"] =37507800, -- [1]
			quest=31304
		},
		[86430] = {
			["icon"] = 17,
			["coord"] =52605150, -- [1]
			 quest=31419,
		},
		[86393] = {
			["desc"] = "不息山的顶峰的云壬的石像边。特效类似于[萨格拉斯的诅咒视界]。",
			["coord"] =44495273, -- [1]
			quest=31417,
		},
		[88723] = {
			["desc"] = "武器架上。",
			["coord"] ={70006380,7120626}, -- [1]
			quest=31421, 
		},
		[86394] = { --山洞
			["desc"] = "深石窟,在洞里第二层,在底层水池中的大石头上.特殊模型.",
			["icon"] = 17,
			["coord"] =52837130, -- [1]
			redir = "TheDeeper"
		},
	},
	[806] = { --翠玉
		[86199] = {
			["desc"] = "npc64272。玉婆婆前青玉战士雕像(坐标：39.2,46.6)。模型类似[流光碧空之歌]",
			["coord"] =39244641, -- [1]
			quest=31307,
		},
		[85777] = {
			["desc"] = "矿洞入口和内部都可能刷新。",
			["icon"] = 17,
			["coord"] =46262952, -- [1]
			redir = "GreenstoneQuarry"
		},
		[86196] = {
			["desc"] = "神龙之心南，说是在河里",
			["coord"] = {
				45306479, -- [1]
				47016757, -- [2]
				45997160, -- [3]
			},
			quest=31402, 
		},
		[86198] = {
			["desc"] = "十雷台的唤魂者之座。特殊模型",
			["coord"] = {
				41251789, -- [2]
				39871430, -- [3]
				41991755,
			},
			quest=31403,
		},
		[85776] = {
			["desc"] = "欧文的许愿井，十雷台西北边。卡拉克西风格模型",
			["coord"] =39360725, -- [1]
			quest=31397,
		},
	},
	[810] = { --螳螂
		[86518] = {
			["desc"] = " 奥苏尔火营，稀有精英隆恩附近的火盆.",
			["coord"] = {66304470,66804800}, -- [1]
			quest=31425, 
		},
	},
	[858] = { --悚然
		[86523] = {
			["desc"] = "闷水壶的冒险码头，海底2个岩石之间的海草上。卡拉克西风格模型",
			["coord"] =56707750, -- [1]
			quest=31434,
		},
		[86519] = {
			["desc"] = "一根底部覆盖着紫色灌木的柱子，比其他的柱子都高。卡拉克西风格模型",
			["coord"] =71883596, -- [1]
			quest=31429,
		},
		[86526] = {
			["desc"] = "暮根沼泽，在沼泽内的一小圈干地，附近有些石头和骨头。",
			["coord"] =54325685, -- [1]
			quest=31437,
		},
		[86524] = {
			["desc"] = "扎尼维斯的白色的岩石，在那个根部。特殊模型。",
			["coord"] =30299083, -- [1]
			 quest=31435, 
		},

		[86522] = {
			["desc"] = "挂在任务怪背后的墙壁上&树根上插着。卡拉克西风格模型",
			["coord"] =66766389, -- [1]
			icon = 28,
			quest=31433, 
		},

		[86529] = {
			["desc"] = "绿名海螺npc65552“拉帕纳微光海螺”对话获得",
			["coord"] =42076305, -- [1]
			quest=31432,
		},
		
		[86527] = {
			["desc"] = "毒水崖 琥珀孵育室门口，藏在灌木丛中比较难发现，",
			["coord"] =28894218, -- [1]
			quest=31438,
		},
		
		[86521] = {
			["desc"] = "恐惧之心西北，靠近水边的三个石头中间。",
			["coord"] =32993012, -- [1]
			quest=31431,
		},
		[86525] = {
			["desc"] = "泥鳞窟，水下洞穴，任务怪背后。",
			["coord"] =25725432, -- [1]
			icon = 28,
			quest=31436,
		},
		[86520] = {
			["desc"] = "找到这样一棵树，附近有3个螳螂人在控制一个巨虫。卡拉克西风格模型。",
			["coord"] =49063030, -- [1]
			quest=31430,
		},
	},
	[857] = { --蛮荒
		[86124] = {
			["desc"] = "站在海滩面朝大海的熊猫人npc德明大叔，就在他屁股后地面插着。简易鱼竿模型",
			["coord"] =50744919, -- [1]
			quest=31409,
		},
	},
	["TheDeeper"] = { -- 崑莱山洞
		[86394] = {
			coord = 24106580,
			["desc"] = "深石窟,在洞里第二层,在底层水池中的大石头上.特殊模型.",
			quest=31413,
		},
    },
	["GreenstoneQuarry"] = { -- 翠玉林山洞
		[85777] = {
			coord = {33107800,44007050,43703850,38704750,32606270,38981480},
			["desc"] = "矿洞入口和内部都可能刷新。",
			quest=31399,
		},
	},
}

local RareElite = { -- 玩具
	[807] = { --四风
		[86565] = {
			["desc"] = "黑蹄",
			["coord"] = {
				35396078, -- [1]
				31856130, -- [2]
				38595790, -- [3]
			},
		},
		[86576] = {
			["desc"] = "撕裂者纳拉克",
			["coord"] = {
				9554611, -- [1]
				12125187, -- [2]
				12285894, -- [3]
				8756392, -- [4]
			},
		},
		[86587] = {
			["desc"] = "納斯拉·斑皮",
			["coord"] =88821809, -- [1]
		},
		[86591] = {
			["desc"] = "波罗波斯",
			["coord"] = 13723878, -- [1]
		},
		[86580] = {
			["desc"] = "塞勒纳",
			["coord"] = {
				58174192, -- [1]
				55932752, -- [2]
				56573616, -- [3]
			},
		},
		[86583] = {
			["desc"] ="萨尔因作战斥候+",
			["coord"] = {
				68285318, -- [1]
				68286287, -- [2]
				72295842, -- [3]
				74385318, -- [4]
			},
		},
		[86569] = {
			["desc"] = "苏里克夏",
			["coord"] =36832647, -- [1]
		},
		[86572]= {
			["desc"] = "虬达",
			["coord"] =18547570, -- [1]
		},
	},
	[809] = { --崑莱
		[86573] = {
			["desc"] ="哈瓦克" ,
			["coord"] =57687348, -- [1]
		},
		[86566] = {
			["desc"] = "考达多罗斯",
			["coord"] =48988009, -- [1]
		},
		[86592] = {
			["desc"] = "斯克里奇",
			["coord"] = {
				44966426, -- [1]
				47556472, -- [2]
				46276198, -- [3]
			},
		},
		[86588] = {
			["desc"] = "漫游者阿波尼+",
			["coord"] =40604210, -- [1]
		},
		[86577] = {
			["desc"] ="斯基提克" ,
			["coord"] =37788072, -- [1]
		},
		[86570] = {
			["desc"] = "波尔金黑拳",
			["coord"] =57534806, -- [1]
		},
		[86584] = {
			["desc"] = "预言者奈松+",
			["coord"] =64631346, -- [1]
		},
		[86581] = {
			["desc"] = "被放逐的诺拉克斯",
			["coord"] =73827716, -- [1]
		},
	},
	[806] = { --翠玉
		[87780] = {
			["desc"] ="马尔泰弱智",
			["coord"] =56013099, -- [1]
		},
		[90078] = {
			["desc"] = "罗布·悲鸣",
			["coord"] =28162155, -- [1]
		},
	},	
	[811] = { --恒春谷
		[86586] = {
			["desc"] = "独眼者摩多",
			["coord"] = {
				39685349, -- [1]
				44114878, -- [2]
				36355946, -- [3]
			},
		},
		[86590] = {
			["desc"] = "迅云艾然",
			["coord"] =41347050, -- [1]
		},
		[86594] = {
			["desc"] ="纳纳斯少校",
			["coord"] =34699203, -- [1]
		},
		[86568] = {
			["desc"] = "利眼约里克",
			["coord"] =86644286, -- [1]
		},
		[86579] = {
			["desc"] = "凋零的卡尔提克",
			["coord"] =14175819, -- [1]
		},
		[86571] = {
			["desc"] = "灵魂窃贼康格",
			["coord"] =14843359, -- [1]
		},
		[86575] = {
			["desc"] ="乌戈拉克斯" ,
			["coord"] =42322402, -- [1]
		},
		[86582] = {
			["desc"] = "撒罕潮猎者",
			["coord"] =61745584, -- [1]
		},
	},
	[858] = { --悚然
		[86589] = {
			["desc"] = "艾利天镜+",
			["coord"] =33812301, -- [1]
		},
		[86578] = {
			["desc"] ="嘎洛克+" ,
			["coord"] =38202888, -- [1]
		},
		[86564] = {
			["desc"] = "黑化者卡尔+",
			["coord"] =71633742, -- [1]
		},
		[86567] = {
			["desc"] ="破坏者哒克" ,
			["coord"] =24822834, -- [1]
		},
		[86585] = {
			["desc"] ="奥米斯格林洛克" ,
			["coord"] = {
				41055971, -- [1]
				38486177, -- [2]
				35246261, -- [3]
				38115846, -- [4]
				36816701, -- [5]
			},
		},
		[86574] = {
			["desc"] = "“利刃”科罗尔+",
			["coord"] =73542277, -- [1]
		},
		[86593] = {
			["desc"] = "聪明的伊克伊克+ 树洞里。",
			["coord"] =54656652, -- [1]
		},
		[86563] = {
			["desc"] = "纳拉什沃登提斯",
			["coord"] =64745765, -- [1]
		},
	},
}

local Achievement1 = { --博学者
	level = 2,
	achieve = true,
	[807] = { --四风
		[684602] = {
			coord = 61203460,
		},
		[685801] = {
			coord = 18803170,
		},
		[685003] = {
			coord = 83102120,
			desc = "似乎有问题 参考30.8 30.5",
		},
		[685601] = {
			coord = 20305590,
		},
		[685602] = {
			coord = 55004720,
		},
		[685603] = {
			coord = 34606400,
		},
	},
	[809] = { --崑莱
		[684604] = {
			coord = 74508350,
		},
		[675401] = { --洞口修正
			coord = 52974655,
			desc = "山洞里(58.29 71.53)",
		},
		[685805] = {
			coord = 63104080,
		},
		[685004] = {
			coord = 45756189,
		},
		[684701] = {
			coord = 50407920,
		},
		[684703] = {
			coord = 71706300,
		},
		[685502] = {
			coord = 43805120,
		},
		[685506] = {
			coord = 67804830,
		},
		[685507] = {
			coord = 41004240,
		},
		[723003] = {
			coord = 44705240,
		},
	},
	[806] = { --翠玉
		[684601] = {
			coord = 66008760,
		},
		[675403] = {
			coord = 42301750,
		},
		[685803] = {
			coord = 35703050,
		},
		[685001] = {
			coord = 26502830,
		},
		[671601] = {
			coord = 67702940,
		},
		[685501] = {
			coord = 47104520,
		},
		[685503] = {
			coord = 55805680,
		},
		[723002] = {
			coord = 37303010,
		},
	},
	[810] = { --螳螂
		[684702] = {
			coord = 65405000,
		},
		[684704] = {
			coord = 84007280,
		},
		[685505] = {
			coord = 37706290,
		},
	},
	[858] = { --悚然
		[685701] = {
			coord = 48303290,
		},
		[685702] = {
			coord = 59905510,
		},
		[685703] = { --洞口修正
			coord = 53601550,
			desc = "洞口"
		},
		[685704] = {
			coord = 35503260,
		},
		[671603] = {
			coord = 67406090,
		},
	},
	[857] = { --蛮荒
		[684603] = {
			coord = 30303860,
		},
		[675402] = {
			coord = 51003170,
		},
		[685002] = { --山洞修正
			coord = 52208600,
			desc = "山洞里面",
		},
		[671604] = {
			coord = 32802950,
		},
		[685504] = {
			coord = 40505660,
			desc = "楼梯往上",
		},
		[723001] = {
			coord = 81401140,
		},
		[685604] = {
			coord = 72203100,
		},
	},
	[811] = { --恒春谷
		[675404] = {
			coord = 40207740,
		},
		[685802] = {
			coord = 52906850,
		},
		[685804] = {
			coord = 26602160,
		},
		[685508] = {
			coord = 68704430,
		},
	},
	[873] = { --客栈
		[671602] = {
			coord = 57721310,
			desc = "山洞",
		},
	},
	["TheAncientPassage"] = { --山洞里
		[671602] = {
			coord = 55131620,
		},
	},
}
--select(4,GetAchievementInfo(6548)) 


local npc = {}
for i = 1,7 do
	npc[i] = GetAchievementCriteriaInfo(6606,i)
end
local PetTamers = {-- =宠物对战
	level = 2,
	icon = 662000,
	[807] = { --四风
		[npc[2]] = {
			coord = 46004360,
			desc = "",
			quest = 31955,
		},
	},
	[809] = { --崑莱
		[npc[4]] = {
			coord = 35807360,
			desc = "",
			quest = 31956,
		},
		["土灵"] = {
			coord = 64809360,
			desc = "",
			quest = 32441,
			icon = 793600,
		},
	},
	[811] = { --恒春谷
		[npc[7]] = {
			coord = 67604080,
			desc = "",
			quest = 31958,
		},
	},
	[858] = { --悚然
		[npc[6]] = {
			coord = 55003760,
			desc = "",
			quest = 31957,
		},
		["水灵"] = {
			coord = 61208760,
			desc = "",
			quest = 32439,
			icon = 793600,
		},
	},
	[806] = {--翠玉
		[npc[1]] = {
			coord = 48005400,
			desc = "",
			quest = 31953,
		},
		["风灵"] = {
			coord = 28803600,
			desc = "",
			quest = 32440,
			icon = 793600,
		},
	},
	[857] = {--蛮荒
		[npc[3]] = {
			coord = 62204580,
			desc = "",
			quest = 31954,
		},
	},
	[810] = {--螳螂
		[npc[5]] = {
			coord = 36205220,
			desc = "",
			quest = 31991,
		},
		["火灵"] = {
			coord = 57004220,
			desc = "",
			quest = 32434,
			icon = 793600,
		},
	},
}

local Event = CreateFrame("Frame")
Event:RegisterEvent("PLAYER_LOGIN")
Event:SetScript("OnEvent",function(self,event)
	if true then tinsert(ns.db,RichOfPandaria) end
	if true then tinsert(ns.db,LostAndFound) end
	if true then tinsert(ns.db,RareElite) end
	if not select(4,GetAchievementInfo(6548))  then tinsert(ns.db,Achievement1) end
	if true then tinsert(ns.db,PetTamers) end
	for _,group in pairs(ns.db) do
		if group.achieve then
			for _,ids in pairs(group) do
				if type(ids)== "table" then
					for id,info in pairs(ids) do
						info._name = GetAchievementCriteriaInfo(ParseAchieveid(id))
					end
				end
			end
		end
	end
end)

---------------------
----以下不要修改----
---------------------

ns.mapscale = {
	[806] = { --翠玉林
		0.4500933135463645, 
		0.4499587726878093,
		0.4705534206845249,
		0.2926539118731085,
	},
	[873] = { --客栈
		0.115611615263164, 
		0.1156116127819066, 
		0.5117761936403419,
		0.4823858033307175,
	},
	[811] = { --恒春谷
		0.1632799786118908, 
		0.1631456303065864,
		0.4042208283541151,
		0.4574103326724259, 
	},
	[858] = { --悚然
		0.3449545330796179,
		0.3450220518230708,
		0.1684327024909846, 
		0.5087710454828307, 
	},
	[857] = { --蛮荒
		0.3021214165847387, 
		0.3021208503206008,
		0.3741431304475718,
		0.6564074552525323,
	},
	[807] = { --四风
		0.2529764589708267, 
		0.2529761597573952,
		0.3914645989001039, 
		0.5397887140930443, 
	},
	[810] = { --螳螂
		0.370199006437579,
		0.3701986912220293,
		0.1078738485208499,
		0.2050388951142818,
	},
	[809] = { --崑莱
		0.4033651894330519,
		0.4034319729651503,
		0.2522206005004135,
		0.1025192009316251, 
	}
}

ns.Icons = {
	[1] =  "Interface\\Minimap\\Tracking\\Auctioneer",
	[2] =  "Interface\\Minimap\\Tracking\\Banker",
	[3] =  "Interface\\Minimap\\Tracking\\BattleMaster",
	[4] =  "Interface\\Minimap\\Tracking\\FlightMaster",
	[5] =  "Interface\\Minimap\\Tracking\\Innkeeper",
	[6] =  "Interface\\Minimap\\Tracking\\Mailbox",
	[7] =  "Interface\\Minimap\\Tracking\\Repair",
	[8] =  "Interface\\Minimap\\Tracking\\StableMaster",
	[9] =  "Interface\\Minimap\\Tracking\\Class",
	[10] =  "Interface\\Minimap\\Tracking\\Profession",
	[11] =  "Interface\\Minimap\\Tracking\\TrivialQuests",
	[12] =  "Interface\\Minimap\\Tracking\\Ammunition",
	[13] =  "Interface\\Minimap\\Tracking\\Food",
	[14] =  "Interface\\Minimap\\Tracking\\Poisons",
	[15] =  "Interface\\Minimap\\Tracking\\Reagents",
	[16] =  "Interface\\PVPFrame\\PVP-ArenaPoints-Icon",
	[17] =  "Interface\\Icons\\Spell_Arcane_PortalDalaran",
	[18] =  "Interface\\Icons\\Tracking_WildPet",
}

wMapNotes.lua
local _,ns = ...
local size = 16 --图标大小

local db,icon_cache = ns.db,ns.Icons

local icons = {}
local index = 0

--/run print(IsQuestFlaggedCompleted(31415))

local GetCurrentMapAreaID,IsQuestFlaggedCompleted,GetAchievementInfo,GetAchievementCriteriaInfo = GetCurrentMapAreaID,IsQuestFlaggedCompleted,GetAchievementInfo,GetAchievementCriteriaInfo
local GetMapInfo,GetItemIcon,GetAchievementLink= GetMapInfo,GetItemIcon,GetAchievementLink
local WorldMapButton = WorldMapButton
local floor,pairs,type,select = floor,pairs,type,select

local function ParseAchieveid(id,num)
	if num == 1 then
		return floor(id/100)
	else
		return floor(id/100),id%100
	end
end

local cave_cache = {}
local function GetCaveInfo(group,name)
	if not cave_cache[name] then
		local item = nil
		for id,info in pairs(db[group][name]) do
			if not(info.quest and IsQuestFlaggedCompleted(info.quest) == 1) then
				item = id
				break
			end
		end
		cave_cache[name] =  item
	end
	return cave_cache[name]
end

local function GetPinFrame()
	index = index + 1
	if #icons < index then
		local icon = CreateFrame("BUTTON",nil,WorldMapButton)
		icon:SetID(index)
		icon:SetSize(size,size)
		icon:SetFrameStrata("HIGH")

		icon.texture = icon:CreateTexture(nil,"ARTWORK")
		icon.texture:SetAllPoints(icon)
		icon.texture:Show()
		
		icon:SetScript("OnEnter",function(self)
			local info = db[self.group][self.mapid][self.name]
			local name = info.redir and GetCaveInfo(self.group,info.redir) or self.name
			GameTooltip:SetOwner(self,"ANCHOR_RIGHT")
			if type(name) == "number" then
				if name>100000 then
					GameTooltip:AddLine(self._name)
					self.link = GetAchievementLink(ParseAchieveid(name))
				else
					GameTooltip:SetItemByID(name)
					_,self.link = GameTooltip:GetItem()
				end
			else
				GameTooltip:AddLine(name)
				self.link = nil
			end
			GameTooltip:AddLine(info.desc)
			GameTooltip:Show()
		end)
		icon:SetScript("OnLeave",function(self)
			GameTooltip:Hide()
		end)
		icon:RegisterForClicks("LeftButtonUp")
		icon:SetScript("OnClick",function(self,button)
			if button == "LeftButton" and self.link then
				DEFAULT_CHAT_FRAME:AddMessage(self.link)
			end
		end)
		tinsert(icons,icon)
	end
	return icons[index]
end

--local Scale = ns.mapscale
local function SetMapPin(group,mapid,coord,name,texture,isCompleted,flag)
	local x,y = floor(coord/10000)/10000,(coord%10000)/10000 --0<x,y<1
		local icon = GetPinFrame()
		icon.mapid = mapid
		icon.name = name
		icon.group = group
		icon:ClearAllPoints()
		if flag then --上一级显示
			local scale = ns.mapscale[mapid] --判断空值
			if not scale then index = index - 1 return end
			icon:SetPoint("CENTER",WorldMapButton,"TOPLEFT",(scale[3]+x*scale[1])*WorldMapButton:GetWidth(),-(scale[4]+y*scale[2])*WorldMapButton:GetHeight())
		else
			icon:SetPoint("CENTER",WorldMapButton,"TOPLEFT",x*WorldMapButton:GetWidth(),-y*WorldMapButton:GetHeight())
		end
		icon.texture:SetTexture(texture)
		if isCompleted then
			icon:SetAlpha(0.5)
		else
			icon:SetAlpha(1)
		end
		icon:Show()
end

local cache_quest = {}
local function GetQuestCompleted(id,force)
	if cache_quest[id] or force then
		cache_quest[id] = IsQuestFlaggedCompleted(id)
	end
	return cache_quest[id]
end

local cache_achieve = {}
local function GetAchieveCompleted(id,force)
	if cache_achieve[id] or force then
		cache_achieve[id] = GetAchievementCriteriaInfo(ParseAchieveid(id))
	end
	return cache_achieve[id]
end

local function GetQuestAndAchieveCompleted(id)
	if type(id) == "number" then
		if id>100000 then 
			return GetAchieveCompleted(id,true)
		else
			return GetQuestCompleted(id,true)
		end
	end
end

local function GetIconCache(id)
	if not icon_cache[id] and type(id) == "number" then
		if id>100000 then
			icon_cache[id] = select(10,GetAchievementInfo(ParseAchieveid(id,1)))
		else
			icon_cache[id] =  GetItemIcon(id)
		end
	end
	return icon_cache[id]
end

local function ImportDate(infos,group,mapid,defaulticon,flag)
	for name,info in pairs(infos) do
		local texture = GetIconCache(info.icon or name) or defaulticon
		local isCompleted = GetQuestAndAchieveCompleted(info.quest or name) or (info.redir and not GetCaveInfo(group,info.redir))
		if type(info.coord) == "number" then 
			SetMapPin(group,mapid,info.coord,name,texture,isCompleted,flag)
		else
			for i,coord in pairs(info.coord) do
				SetMapPin(group,mapid,coord,name,texture,isCompleted,flag)
			end
		end
	end
end

local oldmapid
local Event = CreateFrame("Frame")
Event:RegisterEvent("WORLD_MAP_UPDATE")

Event:SetScript("OnEvent",function(self,event,...)
	if not WorldMapFrame:IsShown() then return end
	local mapid = select(5,GetMapInfo()) or GetCurrentMapAreaID()
	if mapid== oldmapid then return end
	oldmapid =mapid
	index = 0
--依次显示不同组
	for i,group in pairs(ns.db) do
		defaulticon = GetIconCache(group.icon or 1)
		--先导入对应的组
		if group[mapid] then
			ImportDate(group[mapid],i,mapid,defaulticon,false)
		end
		--导入上级的组
		if mapid == 862 and group.level == 2 then
			for _mapid,t  in pairs(group) do
				if type(t) == "table" then
					ImportDate(t,i,_mapid,defaulticon,true)
				end
			end
		end
	end
	for i = index + 1 ,#icons do
		icons[i]:Hide()
	end
end)
--local savedb = {}
SLASH_MAPNOTES1 = "/wmn"
SlashCmdList["MAPNOTES"] = function(cmd)
	local name1,_,_,_,name2 = GetMapInfo()
	DEFAULT_CHAT_FRAME:AddMessage("当前区域id = "..GetCurrentMapAreaID().." 当前区域名称 = "..(name2 or name1))
	local x,y = GetPlayerMapPosition("player")
	DEFAULT_CHAT_FRAME:AddMessage(format("玩家当前坐标:x = %.4f y = %.4f ",x*100,y*100))
	--DEFAULT_CHAT_FRAME:AddMessage("玩家位置坐标:"
	-- for mapid,t1 in pairs(db) do
		-- if not savedb[mapid] then savedb[mapid] = {} end
		-- for coord,t2 in pairs(t1) do
			-- id = t2.item or t2.npc or t2.title or "未知"
			-- if not savedb[mapid][id] then savedb[mapid][id] = {} end
			-- if not savedb[mapid][id].coord then savedb[mapid][id].coord = {} end
			-- tinsert(savedb[mapid][id].coord ,coord)
			-- if t2.desc then savedb[mapid][id].desc = t2.desc end
			-- savedb[mapid][id].icon = t2.icon 
		-- end
	-- end
	-- wMapNotesDB = savedb
	-- print("转换数据成功,重载后保存")
end